tests/unit/test_platform.cpp

```diff
@@ -170,7 +171,14 @@ OT_TOOL_WEAK otRadioFrame *otPlatRadioGetTransmitBuffer(otInstance *) { return n
 
 OT_TOOL_WEAK int8_t otPlatRadioGetRssi(otInstance *) { return 0; }
 
-OT_TOOL_WEAK otRadioCaps otPlatRadioGetCaps(otInstance *) { return OT_RADIO_CAPS_NONE; }
+OT_TOOL_WEAK otRadioCaps otPlatRadioGetCaps(otInstance *)
+{
+    return
+#if !OPENTHREAD_CONFIG_MAC_SOFTWARE_CSMA_BACKOFF_ENABLE
+        OT_RADIO_CAPS_CSMA_BACKOFF |
+#endif
```

Comment:
Same here! We should not claim that the radio provides this capability when it actually does not.

> This commit also changes some test platform's capability declaration to pass the assertion.

This PR introduces a new `OT_ASSERT()` that checks `OPENTHREAD_CONFIG_MAC_SOFTWARE_CSMA_BACKOFF_ENABLE` and requires it to be defined and used on an `MTD/FTD` build, which, as I pointed out earlier, is unnecessary.

This caused some tests to fail.

* The correct approach would be to ensure all test platform implementations define `OPENTHREAD_CONFIG_MAC_SOFTWARE_CSMA_BACKOFF_ENABLE` and set it to one.
* Modifying `otPlatRadioGetCaps` simply to pass the assertion strikes me as a poor solution.